### PR TITLE
Block Fetch Demo

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/ChainFragment.hs [hunk 1]
```diff
@@ -379,6 +380,16 @@ dropWhileNewest p c@(c' :> b)
                   | p b       = dropWhileNewest p c'
                   | otherwise = c
 
+takeWhileOldest :: HasHeader block
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/ChainFragment.hs]: Could you add the following property test to `Test.ChainFragment`?

```haskell

prop_takeWhileOldest :: (Block -> Bool) -> TestBlockChainFragment -> Bool
prop_takeWhileOldest p (TestBlockChainFragment chain) =
    CF.takeWhileOldest p chain
 == (CF.fromOldestFirst . L.takeWhile p . CF.toOldestFirst) chain
```

File: ouroboros-network/demo/chain-sync.hs [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+    }
+
+--
+-- Utils
+--
+
+genesisChainFragment :: AF.AnchoredFragment BlockHeader
+genesisChainFragment = AF.Empty (Point 0 GenesisHash)
+
+shiftAnchoredFragment :: HasHeader block
```
Comment (mrBliss) [inline ouroboros-network/demo/chain-sync.hs]: There is also `AnchoredFragment.anchorNewest`, but that works with `Chain`s. Maybe using `Chain`s here might make things easier, or we could provide a variant of `anchorNewest` that takes an `AnchoredFragment` as input.

Also: the alignment of the arrows is off.

File: ouroboros-network/demo/chain-sync.hs [hunk 3]
```diff
@@ -101,10 +107,17 @@ mkLocalSocketAddrInfo socketPath =
...
+
 defaultLocalSocketAddrInfo :: Socket.AddrInfo
 defaultLocalSocketAddrInfo = 
-    mkLocalSocketAddrInfo "./demo-chain-sync.sock"
+    mkLocalSocketAddrInfo defaultLocalSocketAddrPath
 
+rmDefaultLocalSocketAddr :: IO ()
+rmDefaultLocalSocketAddr = do
+  b <- doesFileExist defaultLocalSocketAddrPath
+  when b (removeFile defaultLocalSocketAddrPath)
```
Comment (dcoutts) [inline ouroboros-network/demo/chain-sync.hs]: It'd probably be more convenient to delete the socket in all the server cases, not just the one where the name is the default one. Not a big deal.

File: ouroboros-network/demo/chain-sync.hs [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+import System.Exit
+import System.Random
+import System.Random.SplitMix
+
+import qualified Codec.Serialise as CBOR
+import           Codec.SerialiseTerm
+
+import qualified Network.Socket as Socket
+
+--import Ouroboros.Network.NodeToNode
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply ouroboros-network/demo/chain-sync.hs]: Ideally we should indeed tidy this up. The bigger issue is actually elsewhere in the demo where there was a bit I didn't get working.

File: ouroboros-network/src/Ouroboros/Network/ChainFragment.hs (hunk 1 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ChainFragment.hs]: Done

File: ouroboros-network/demo/chain-sync.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/demo/chain-sync.hs]: I'll leave `shiftAnchoredFragment` as is for now, I'll only fix the alignment. `anchorNewest` based on `AnchoredFragment` might be useful for testing the volitile db.

File: ouroboros-network/demo/chain-sync.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/demo/chain-sync.hs]: Right; fixed.

**Pull request merged**